scripts/vault-push-approles.nix

```diff
@@ -0,0 +1,251 @@
+# Generate and push approles to vault
+{ writeShellScriptBin, jq, vault, coreutils, lib }:
+# Inputs: a flake with `nixosConfigurations`
+
+# Usage:
+# apps.x86_64-linux.vault-push-approles = { type = "app"; program = "${pkgs.vault-push-approles self}/bin/vault-push-approles"; }
+
+{ nixosConfigurations ? { }, ... }: rec {
+  # Overrideable functions
+  # Usage examples:
+  # pkgs.vault-push-approles self { approleCapabilities.aquarius-albali-borgbackup = [ "read" "write" ]; }
+  /* pkgs.vault-push-approles self (final: prev: {
+       approleCapabilitiesFor = { approleName, namespace, ... }@params:
+         if namespace == "albali" then [ "read" "write" ] else prev.approleCapabilitiesFor params;
+     })
+  */
+  # pkgs.vault-push-approles { } { extraApproles = [ { ... } ] }
+
+  # `final` contains fixed-point functions after applying user-supplied overrides
+  overrideable = final: {
+
+    # Approles to upload in addition to the ones generated from
+    # vault-secrets' NixOS module definitions. Must contain all the
+    # options from top-level vault-secrets option and all the options
+    # from vault-secrets.secrets.<name> submodule, as well as
+    # "approleName" attribute
+    extraApproles = [ ];
+
+    # Render an attrset into a JSON file
+    renderJSON = name: content:
+      builtins.toFile "${name}.json" (builtins.toJSON content);
+
+    # Default approle parameters
+    approleParams = {
+      secret_id_ttl = "";
+      token_num_uses = 0;
+      token_ttl = "20m";
+      token_max_ttl = "30m";
+      secret_id_num_uses = 0;
+    };
+
+    # Generate an approle parameters attrset based on its name and other
+    # options from its secret definition
+    mkApprole = { approleName, ... }:
+      (final.approleParams // { token_policies = [ approleName ]; });
+
+    # Create a JSON (HCL) file with approle parameters in it from its secret definition
+    renderApprole = { approleName, ... }@params:
+      final.renderJSON "approle-${approleName}" (final.mkApprole params);
+
+    # An attrset mapping `approleName`s to capabilities required by those approles
+    approleCapabilities = { };
+
+    # Get capabilities for the given secret definition
+    approleCapabilitiesFor = { approleName, ... }:
+      final.approleCapabilities.${approleName} or [ "read" ];
+
+    # Generate an approle policy from its secret definition
+    mkPolicy = { approleName, name, vaultPathPrefix, namespace, ... }@params:
+      let
+        splitPrefix = builtins.filter builtins.isString
+          (builtins.split "/" vaultPathPrefix);
+
+        insertAt = lst: index: value:
+          (lib.lists.take index lst) ++ [ value ] ++ (lib.lists.drop index lst);
+
+        makePrefix = value:
+          builtins.concatStringsSep "/" (insertAt splitPrefix 1 value);
```

Comment:
Note that it assumes that mountpoint path consists of one component, but in our usecase it's true, so it's fine

Reply:
Should I fix this then somehow?

Reply:
I think it's not important, if we need it we can implement it later